### ISSUE TITLE
[css-values-5] Proposal for a new `match()` CSS Function

### ISSUE THREAD

**Issue by brandonmcconnell** [w3c/csswg-drafts#10594]
## Abstract

The `match()` function is proposed to enhance the conditional application of styles in CSS by checking if the current element matches a specified selector. This function complements the [recently accepted](https://github.com/w3c/csswg-drafts/issues/10064#issuecomment-2165157958) new [`if()` function](https://github.com/w3c/csswg-drafts/issues/10064#issue-2182499866), providing a more semantic and readable way to handle conditional logic within CSS properties.

## Motivation

The current method of conditionally applying styles in CSS often involves the use of custom properties (variables) and mathematical operations, which can become complex and less readable. The introduction of `if()` allows for clearer boolean logic handling. However, the need to constantly check selector matches remains manually verbose. The `match()` function aims to address this by allowing a direct approach to conditionally altering styles based on selector matches.

## Syntax

```css
match(selector)
```

- **selector**: A CSS selector to check against the current element.

## Usage

The `match()` function can be used inside any CSS property value expression where conditional styles are necessary based on whether the current element matches a specified selector.

### Example 1: Basic example

```postcss
.item {
  --symbol: if(match(.favorite) ? "🧡" : "🩶");
  &::before {
    content: var(--symbol);
  }
}
```

In this example, the `--symbol` custom property is conditionally set to different emoji characters based on whether the `.item` element has the `.favorite` class.

### Example 2.1: Without `if()` and `match()`

```postcss
.item {
  --symbol: "🩶";
  &::before {
    content: var(--symbol);
  }
  &.favorite {
    --symbol: "🧡";
  }
}
```

This traditional approach involves redundant code as it sets the same property (`--symbol`) in multiple places.

### Example 2.2: Using `if()` and `match()`

```postcss
.item {
  --symbol: if(match(.favorite) ? "🧡" : "🩶");
  &::before {
    content: var(--symbol);
  }
}
```

This approach reduces redundancy and centralizes the logic for setting the `--symbol` property.

## A case for boolean expressions outside `if()`

I've opened a separate proposal focused on supporting a new "boolean" custom property type, which could hypothetically be valid and type-safe even outside of `if()`. If that's accepted, an example using a math-based switch like the one below could be simplified to use boolean logic.

### Example 3.1: Using a math-based switch

This example uses a math-based switch to zero out values when a condition is not met.

```postcss
label {
  --selected: 0;
  &:has(:checked) { --selected: 1; }
  grid-template-columns: auto calc(var(--selected) * 20px);
  svg { opacity: var(--selected); }
}
```

### Example 3.2: Using `match()` and `if()`

This moves the conditions inline, but they're a bit complex, so it's not ideal that we have to repeat the conditions in both values.

```postcss
label {
  grid-template-columns: auto if(match(:has(:checked)) ? 20px : 0);
  svg { opacity: if(match(:has(:checked)) ? 1 : 0); }
}
```

### Example 3.3: Using `match()` and `if()`

Using reusable boolean values can offload some of this complexity.

```postcss
label {
  --selected: match(:has(:checked));
  grid-template-columns: auto if(var(--selected) ? 20px : 0);
  svg { opacity: if(var(--selected) ? 1 : 0); }
}
```

**Comment by romainmenke** [comment]
Shouldn't `if` inherit like `light-dark` with custom properties?

see : https://codepen.io/romainmenke/pen/wvLGwyb

```html
<div id=a>
  hello
  <div id=b>
    world
  </div>
</div>
```

```css
:root {
  --fg: light-dark(black, white);
  --bg: light-dark(white, black);
}

#a {
  color-scheme: dark;
  color: var(--fg);
  background-color: var(--bg);
}

#b {
  color-scheme: light;
  color: var(--fg);
  background-color: var(--bg);
}
```

In this example `--fg` and `--bg` are not "evaluated" to specific colors based on the color scheme of `:root`. Instead the `light-dark()` functions are inherited and evaluated on specific elements against their respective color scheme.

So in your first examples `matches()` would never work because those pseudo elements would never have the `.favorite` class. Right?

I think you would need to write:

```css
.item {
  --symbol: if(matches(.favorite::before) ? "🧡" : "🩶");
  &::before {
    content: var(--symbol);
  }
}
```

I think authors will often be surprised by such behavior because you need to consider the outer selector, how values cascade and inherit and how both will affect the inner selector.

Is the added complexity worth it, considering that we can already style elements by matched selector?

**Comment by josephrocca** [comment]
Please forgive that this question is somewhat tangential to this proposal, and feel free to hide this comment if it's considered off topic (i.e. should be its own separate issue). It's about both `if` and `matches`, in relation to helpers like `light-dark()`. Please feel free to just read the TL;DR at the end.

### (Context) In search of `light-dark()`-level simplicity for hovered elements
I was about to create an issue in this repo asking whether something like `light-dark()` is possible, but for `:hover`ed elements. I *love* the simplicity of `light-dark()`, and I've always wanted to be able to define hover colors in inline styles. Something like this:

```html
<div style="color:hover-else(blue, black)">this text turns blue when hovered</div>
```
While researching before creating the issue, I came across the `if`/`matches` proposals. If I understand correctly, it would look something like this:
```html
<div style="color:if(matches(:hover) blue ? black)">this text turns blue when hovered</div>
```

### (Question) Can this proposal be adapted to reduce nesting for simple cases, or should it be a separate proposal?
For a common case like hover-based color changes, the degree of nesting and syntax required to achieve this using `if`/`matches` seems a bit too far. I guess I see a few possibilities here:

1. A separate proposal for a helper function like `hover-else`.
2. Cover the common "if <selector>, valueX, else valueY" case without requiring two levels of nesting, at least. Something like `color:cond(:hover ? blue : black)`.
3. Allow `matches` to return values in place of `true` and `false` if they're given as two extra arguments - e.g. `color:matches(:hover, blue, black)`.
4. Forsake the "simple things should be simple" principle and just make devs write `color:if(matches(:hover) blue ? black)`.

My preference is 3, and I'm wondering what others think about this? I.e.:

```html
<div style="color:matches(:hover, blue, black)">this text turns blue when hovered</div>
```
**TL;DR**: Can we allow `matches` to return values in place of `true` and `false` if they're given as two extra arguments - e.g. `color:matches(:hover, blue, black)` would return `blue` in place of `true` and `black` in place of `false`?
